src/tip-opengov.ts

```diff
@@ -1,154 +1,132 @@
-import { ISubmittableResult } from "@polkadot/types/types";
-import type { BN } from "@polkadot/util";
+import {
+  GovernanceOrigin,
+  PolkadotRuntimeOriginCaller,
+  PreimagesBounded,
+  TraitsScheduleDispatchTime,
+  WestendRuntimeGovernanceOriginsPalletCustomOriginsOrigin,
+  WestendRuntimeOriginCaller,
+} from "@polkadot-api/descriptors";
+import { ss58Address } from "@polkadot-labs/hdkd-helpers";
+import { getDescriptor } from "#src/chain-config";
+import { Binary, PolkadotClient, TxPromise } from "polkadot-api";
 import { Probot } from "probot";
 
 import { Polkassembly } from "./polkassembly/polkassembly";
-import { ContributorAccount, OpenGovTrack, State, TipRequest, TipResult } from "./types";
-import { encodeProposal, formatReason, getReferendumId, tipSizeToOpenGovTrack } from "./util";
+import { OpenGovTrack, State, TipNetwork, TipRequest, TipResult } from "./types";
+import { encodeProposal, formatReason, tipSizeToOpenGovTrack } from "./util";
 
-type ExtrinsicResult = { success: true; blockHash: string } | { success: false; errorMessage: string };
-
-export function tipOpenGovReferendumExtrinsic(opts: { api: ApiPromise; tipRequest: TipRequest }):
+export async function tipOpenGovReferendumExtrinsic(opts: { client: PolkadotClient; tipRequest: TipRequest }): Promise<
   | Exclude<TipResult, { success: true }>
   | {
       success: true;
-      referendumExtrinsic: SubmittableExtrinsic<"promise">;
+      referendumExtrinsic: { signAndSubmit: TxPromise<"promise"> };
       proposalByteSize: number;
-      encodedProposal: string;
-      track: { track: OpenGovTrack; value: BN };
-    } {
-  const { api, tipRequest } = opts;
+      encodedProposal: Binary;
+      track: { track: OpenGovTrack; value: bigint };
+    }
+> {
+  const { client, tipRequest } = opts;
   const track = tipSizeToOpenGovTrack(tipRequest);
   if ("error" in track) {
     return { success: false, errorMessage: track.error };
   }
 
-  const encodeProposalResult = encodeProposal(api, tipRequest);
+  const encodeProposalResult = await encodeProposal(client, tipRequest);
   if ("success" in encodeProposalResult) {
     return encodeProposalResult;
   }
   const { encodedProposal, proposalByteSize } = encodeProposalResult;
 
-  const referendumExtrinsic = api.tx.referenda.submit(
-    // TODO: There should be a way to set those types properly.
-    { Origins: track.track.trackName } as never,
-    { Inline: encodedProposal },
-    { after: 10 } as never,
-  );
+  const proposal = PreimagesBounded.Inline(encodedProposal);
+
+  const enactMoment = TraitsScheduleDispatchTime.After(10);
+
+  let referendumExtrinsic: { signAndSubmit: TxPromise<"promise"> };
+  const network: TipNetwork = tipRequest.contributor.account.network;
+  if (network === "westend" || network === "localwestend" || network === "rococo" || network === "localrococo") {
+    const api = client.getTypedApi(getDescriptor(network));
+    const proposalOrigin = WestendRuntimeOriginCaller.Origins(
```

Comment:
So, this `WestendRuntimeOriginCaller` is gonna work on Rococo?

And `PolkadotRuntimeOriginCaller` is gonna work on Kusama?

Reply:
Yeah, so what papi does, is it creates a checksum for every type, and normalizes the types by that checksum. If a known type results in different checksums in different chains, the resulting type will be prefixed by a name of one of the chains. 

The mapping from checksums to names is statically generated, here are the OriginCaller variants:
https://github.com/polkadot-api/polkadot-api/blob/ce606392f5fa85294fb9efa8dd1c4fd635147da8/packages/codegen/src/known-types.ts#L1302-L1307
https://github.com/polkadot-api/polkadot-api/blob/ce606392f5fa85294fb9efa8dd1c4fd635147da8/packages/codegen/src/known-types.ts#L720-L725

Reply:
I see.

Do you know if it can be a problem that `WestendRuntimeOriginCaller` has `chains: "rococo, westend"`, but `PolkadotRuntimeOriginCaller` has `chains: "polkadot"` (and no `kusama`)?

Reply:
Not from the types perspective. In the end, it's just a name in TS, which corresponds to a specific checksum. 
The type of this checksum is present and expected by Kusama's metadata. The name isn't present there.